mitiq/folding_cirq.py

```diff
@@ -1,76 +1,197 @@
+            if num_folded == num_to_fold:
+                return folded
 
 
-def fold_gates(circuit: Circuit, sub_indices: List[int]) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit
-    determined by sub_indices."""
+def fold_gates_at_random(circuit: Circuit, stretch: float, **kwargs) -> Circuit:
+    """Returns a folded circuit by applying the map G -> G G^dag G to a random subset of gates in the input circuit.
 
-    out = Circuit()
-    for j, gate in enumerate(circuit):
-        out += gate
-        if j in sub_indices:
-            out += inverse(gate)
-            out += gate
-    return out
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
 
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
 
-def local_folding(circuit: Circuit, stretch: float, stretcher: Callable=left_stretcher) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
 
-    if not (stretch >= 1):
-        raise ValueError("The stretch factor must be a real number >= 1.")
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
 
-    if stretch <= 3:
-        return stretcher(circuit, stretch)
-    else:
-        # recursive iterations for stretch > 3
-        _ = local_folding(circuit, 3, stretcher)
-        return local_folding(_, stretch / 3, stretcher)
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
+    if "seed" in kwargs.keys():
+        np.random.seed(kwargs.get("seed"))
 
-# ==================================================
-# Circuit level folding
-# ==================================================
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
 
+    for _ in range(num_to_fold):
+        # TODO: This allows for gates to be folded more than once, and folded gates to be folded.
+        #  Should this be allowed?
+        moment_index = np.random.choice(len(folded))
+        gate_index = np.random.choice(len(folded[moment_index]))
+        # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
+        #  ==> Possible fix: Have fold_gate_at_index_in_moment(...) modify the circuit in place.
+        folded = fold_gate_at_index_in_moment(folded, moment_index, gate_index)
 
+    return folded
+
+
+def fold_local(circuit: Circuit, stretch: float, fold_method: Callable = fold_gates_from_left, **kwargs) -> Circuit:
```

Comment:
You can add the type signature into the Callable type as follows
```suggestion
def fold_local(circuit: Circuit, stretch: float, fold_method: Callable[[Circuit, float], Circuit] = fold_gates_from_left, **kwargs) -> Circuit:
```

Reply:
Also, are the kwargs here used anywhere? I didn't catch it but may have missed it. It is one possible way to extend thing, but if we don't need this extension yet I'd recommend removing the `**kwargs`

Reply:
> You can add the type signature into the Callable type as follows

I didn't know this -- nice!

Reply:
> Also, are the kwargs here used anywhere? I didn't catch it but may have missed it. It is one possible way to extend thing, but if we don't need this extension yet I'd recommend removing the `**kwargs`

The `fold_method` could have `kwargs`, so I included `kwargs` in `fold_local`. For example, `fold_gates_at_random` has `seed` as a kwarg (for testing).

I admit the only keyword arguments there are now is a random seed for testing, but it could be good to allow for this. We might come up with some wild `fold_method`s...

Reply:
true, but **kwargs can be risky as it hides function signatures. My suggestion would be to explicitly make it `seed` as the argument for now. If we end up with lots of different args then we can consider a refactor later